src/ssl.c

```diff
@@ -49571,87 +49570,160 @@ static int get_dn_attr_by_nid(int n, const char** buf)
-        for (i = 0; i < count; i++) {
+    for (i = 0; i < count; i++) {
+        /* reverse name order for RFC2253 and DN_REV */
+        if ((flags & XN_FLAG_RFC2253) || (flags & XN_FLAG_DN_REV)) {
             ne = wolfSSL_X509_NAME_get_entry(name, count - i - 1);
-            if (ne == NULL)
-                return WOLFSSL_FAILURE;
-
-            str = wolfSSL_X509_NAME_ENTRY_get_data(ne);
-            if (str == NULL)
-                return WOLFSSL_FAILURE;
+        } else {
+            ne = wolfSSL_X509_NAME_get_entry(name, i);
+        }
+        if (ne == NULL)
+            return WOLFSSL_FAILURE;
 
-            len = get_dn_attr_by_nid(ne->nid, &buf);
-            if (len == 0 || buf == NULL)
-                return WOLFSSL_FAILURE;
+        str = wolfSSL_X509_NAME_ENTRY_get_data(ne);
+        if (str == NULL)
+            return WOLFSSL_FAILURE;
 
-            tmpSz = str->length + len + 2; /* + 2 for '=' and comma */
-            if (tmpSz > ASN_NAME_MAX) {
-                WOLFSSL_MSG("Size greater than ASN_NAME_MAX");
+        if (flags & XN_FLAG_RFC2253) {
+            /* escape string for RFC 2253, ret sz not counting null term */
+            escapeSz = wolfSSL_EscapeString_RFC2253(str->data,
+                            str->length, escaped, sizeof(escaped));
+            if (escapeSz < 0)
                 return WOLFSSL_FAILURE;
-            }
 
-            if (i < count - 1) {
-                /* tmpSz+1 for last null char */
-                XSNPRINTF(tmp, tmpSz+1, "%s=%s,", buf, str->data);
-                XSTRNCAT(fullName, tmp, tmpSz+1);
-            }
-            else {
-                XSNPRINTF(tmp, tmpSz, "%s=%s", buf, str->data);
-                XSTRNCAT(fullName, tmp, tmpSz-1);
-                tmpSz--; /* Don't include null char in tmpSz */
-            }
-            totalSz += tmpSz;
+            nameStr = escaped;
+            nameStrSz = escapeSz;
         }
-        if (wolfSSL_BIO_write(bio, fullName, totalSz) != totalSz)
+        else {
+            nameStr = str->data;
+            nameStrSz = str->length;
+        }
+
+        /* len is without null terminator */
+        len = get_dn_attr_by_nid(ne->nid, &buf);
+        if (len == 0 || buf == NULL)
             return WOLFSSL_FAILURE;
-        return WOLFSSL_SUCCESS;
-    }
-#else
-    if (flags == XN_FLAG_RFC2253) {
-        if ((name->sz < 3) ||
-            (wolfSSL_BIO_write(bio, name->name + 1, name->sz - 2)
-                                                            != name->sz - 2))
+
+        tmpSz = nameStrSz + len + 3; /* + 3 for '=', comma, and '\0' */
+        tmp = (char*)XMALLOC(tmpSz, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+        if (tmp == NULL) {
             return WOLFSSL_FAILURE;
+        }
+
+        if (i < count - 1) {
+            XSNPRINTF(tmp, tmpSz, "%s=%s,", buf, nameStr);
+            tmpSz = len + nameStrSz + 2; /* 2 for '=', comma */
+        }
+        else {
+            XSNPRINTF(tmp, tmpSz, "%s=%s", buf, nameStr);
```

Comment:
The body of the `else` is the same as the `if`, aside from the comment.

Reply:
The XSNPRINTF of the if puts a comma after the string `%s=%s,` where the else does not `%s=%s` (no comma), so slightly different from each other.